homeassistant/components/tasmota/device_trigger.py

```diff
@@ -0,0 +1,287 @@
+    hass: HomeAssistantType = attr.ib()
+    remove_update_signal: Callable[[], None] = attr.ib()
+    subtype: str = attr.ib()
+    tasmota_trigger: TasmotaTrigger = attr.ib()
+    type: str = attr.ib()
+    trigger_instances: List[TriggerInstance] = attr.ib(factory=list)
+
+    async def add_trigger(self, action, automation_info):
+        """Add Tasmota trigger."""
+        instance = TriggerInstance(action, automation_info, self)
+        self.trigger_instances.append(instance)
+
+        if self.tasmota_trigger is not None:
+            # If we know about the trigger, set it up
+            await instance.async_attach_trigger()
+
+        @callback
+        def async_remove() -> None:
+            """Remove trigger."""
+            if instance not in self.trigger_instances:
+                raise HomeAssistantError("Can't remove trigger twice")
+
+            if instance.remove:
+                instance.remove()
+            self.trigger_instances.remove(instance)
+
+        return async_remove
+
+    def detach_trigger(self):
+        """Remove Tasmota device trigger."""
+        # Mark trigger as unknown
+        self.tasmota_trigger = None
+
+        # Unsubscribe if this trigger is in use
+        for trig in self.trigger_instances:
+            if trig.remove:
+                trig.remove()
+                trig.remove = None
+
+    async def arm_tasmota_trigger(self):
+        """Arm Tasmota trigger: subscribe to MQTT topics and fire events."""
+
+        @callback
+        def _on_trigger():
+            data = {
+                "mac": self.tasmota_trigger.cfg.mac,
+                "source": self.tasmota_trigger.cfg.source,
+                "event": self.tasmota_trigger.cfg.event,
+            }
+            self.hass.bus.async_fire(
+                TASMOTA_EVENT,
+                data,
+            )
+
+        self.tasmota_trigger.set_on_trigger_callback(_on_trigger)
+        await self.tasmota_trigger.subscribe_topics()
+
+    async def set_tasmota_trigger(self, tasmota_trigger, remove_update_signal):
+        """Set Tasmota trigger."""
+        await self.update_tasmota_trigger(tasmota_trigger.cfg, remove_update_signal)
+        self.tasmota_trigger = tasmota_trigger
+
+        for trig in self.trigger_instances:
+            await trig.async_attach_trigger()
+
+    async def update_tasmota_trigger(self, tasmota_trigger_cfg, remove_update_signal):
+        """Update Tasmota trigger."""
+        self.remove_update_signal = remove_update_signal
+        self.type = tasmota_trigger_cfg.type
+        self.subtype = tasmota_trigger_cfg.subtype
+
+
+async def async_setup_trigger(hass, tasmota_trigger, config_entry, discovery_hash):
+    """Set up a discovered Tasmota device trigger."""
+    discovery_id = tasmota_trigger.cfg.trigger_id
+    remove_update_signal = None
+    _LOGGER.debug(
+        "Discovered trigger with id: %s '%s'", discovery_id, tasmota_trigger.cfg
+    )
+
+    async def discovery_update(trigger_config):
```

Comment:
By defining this method inside this method, it will be defined for each trigger. Can we make this reusable by moving it out and using functools.partial?

Reply:
Hmm, I don't think closure are much worse than using `functools.partial`. Let's keep it like this for now.